File: src/foam/nanos/cron/CronScheduler.java
/**
 * @license
 * Copyright 2017 The FOAM Authors. All Rights Reserved.
 * http://www.apache.org/licenses/LICENSE-2.0
 */

package foam.nanos.cron;

import foam.core.ContextAwareSupport;
import foam.core.Detachable;
import foam.core.FObject;
import foam.dao.AbstractDAO;
import foam.dao.AbstractSink;
import foam.dao.DAO;
import foam.dao.MapDAO;
import foam.mlang.MLang;
import foam.mlang.sink.Min;
import foam.nanos.auth.EnabledAware;
import foam.nanos.logger.Logger;
import foam.nanos.NanoService;
import foam.nanos.script.ScriptStatus;
import foam.nanos.pm.PM;
import java.util.Date;

public class CronScheduler
  extends    ContextAwareSupport
  implements NanoService, Runnable, EnabledAware
{
  protected static long CRON_DELAY = 5000L;

  protected DAO cronDAO_;

  protected boolean enabled_ = true;

  /**
   * Gets the minimum scheduled cron job
   *
   * @return Date of the minimum scheduled cron job
   */
  private Date getMinScheduledTime() {
    Min min = (Min) cronDAO_
      .where(MLang.EQ(Cron.ENABLED, true))
      .select(MLang.MIN(Cron.SCHEDULED_TIME));

    if ( min.getValue().equals(0) ) {
      return null;
    }

    return (Date) min.getValue();
  }

  public void setEnabled(boolean enabled) {
    enabled_ = enabled;
  }

  public boolean getEnabled() {
    return enabled_;
  }

  public void start() {
    cronDAO_ = (DAO) getX().get("cronDAO");

    new Thread(this).start();
  }

  @Override
  public void run() {
    final Logger logger = (Logger) getX().get("logger");

    try {
      while ( true ) {
        if ( getEnabled() ) {
          Date now = new Date();

          cronDAO_.where(
                         MLang.AND(
                                   MLang.LTE(Cron.SCHEDULED_TIME, now),
                                   MLang.EQ(Cron.ENABLED, true),
                                   MLang.IN(Cron.STATUS, new ScriptStatus[] {
                                                          ScriptStatus.UNSCHEDULED,
                                                          ScriptStatus.ERROR,
                                     })
                                   )
                         )
            .orderBy(Cron.SCHEDULED_TIME)
            .select(new AbstractSink() {
                             @Override
                             public void put(Object obj, Detachable sub) {
                               Cron cron = (Cron) ((FObject) obj).fclone();
                               PM pm = new PM(CronScheduler.class, "cron:" + cron.getId());
                               try {
                                 cron.setStatus(ScriptStatus.SCHEDULED);
                                 cronDAO_.put(cron);
                               } catch (Throwable t) {
                                 logger.error(this.getClass(), "Error running Cron Job", cron.getId(), t.getMessage(), t);
                               } finally {
                                 pm.log(getX());
                               }
                             }
                           });
        }
        // Check for new cronjobs every 5 seconds if no current jobs
        // or if their next scheduled execution time is > 5s away
        // Delay at least a little bit to avoid blocking in case of a script error.
        long delay = CRON_DELAY;
        Date minScheduledTime = getMinScheduledTime();
        if( minScheduledTime != null &&
            getEnabled() ) {
          delay = Math.abs(minScheduledTime.getTime() - System.currentTimeMillis());
          delay = Math.min(CRON_DELAY, delay);
          delay = Math.max(500, delay);
        }
        Thread.sleep(delay);
      }
    } catch (Throwable t) {
      logger.error(this.getClass(), t.getMessage());
    }
  }
}

File: src/foam/nanos/cron/CronScheduler.js
/**
 * @license
 * Copyright 2017 The FOAM Authors. All Rights Reserved.
 * http://www.apache.org/licenses/LICENSE-2.0
 */

foam.CLASS({
  package: 'foam.nanos.cron',
  name: 'CronScheduler',

  implements: [
    'foam.core.ContextAgent',
    'foam.nanos.auth.EnabledAware',
    'foam.nanos.NanoService'
  ],

  documentation: ``,

  javaImports: [
    'foam.core.Agency',
    'foam.core.AgencyTimerTask',
    'foam.core.ContextAwareSupport',
    'foam.core.Detachable',
    'foam.core.FObject',
    'foam.dao.AbstractDAO',
    'foam.dao.AbstractSink',
    'foam.dao.DAO',
    'foam.dao.MapDAO',
    'foam.mlang.MLang',
    'foam.mlang.sink.Min',
    'foam.nanos.auth.EnabledAware',
    'foam.nanos.logger.PrefixLogger',
    'foam.nanos.logger.Logger',
    'foam.nanos.NanoService',
    'foam.nanos.script.ScriptStatus',
    'foam.nanos.pm.PM',
    'java.util.Date',
    'java.util.Timer'
  ],

  properties: [
    {
      name: 'cronDelay',
      class: 'Long',
      value: 5000
    },
    {
      name: 'initialTimerDelay',
      class: 'Long',
      value: 60000
    },
    {
      name: 'cronDAO',
      class: 'foam.dao.DAOProperty',
      visibility: 'HIDDEN',
      javaFactory: `
      return (DAO) getX().get("cronDAO");
      `
    },
    {
      name: 'enabled',
      class: 'Boolean',
      value: true
    },
    {
      name: 'timer',
      class: 'Object',
      visibility: 'HIDDEN'
    },
    {
      name: 'logger',
      class: 'FObjectProperty',
      of: 'foam.nanos.logger.Logger',
      visibility: 'HIDDEN',
      javaFactory: `
        return new PrefixLogger(new Object[] {
          this.getClass().getSimpleName()
        }, (Logger) getX().get("logger"));
      `
    }
  ],

  methods: [
    {
      documentation: 'NanoService implementation.',
      name: 'start',
      javaCode: `
      Timer timer = new Timer(this.getClass().getSimpleName());
      setTimer(timer);
      timer.schedule(
        new AgencyTimerTask(getX(), this),
        getInitialTimerDelay());
      `
    },
    {
      documentation: 'Get the minimum scheduled cron job',
      name: 'getMinScheduledTime',
      type: 'DateTime',
      javaCode: `
    Min min = (Min) getCronDAO()
      .where(MLang.EQ(Cron.ENABLED, true))
      .select(MLang.MIN(Cron.SCHEDULED_TIME));

    if ( min.getValue().equals(0) ) {
      return null;
    }

    return (Date) min.getValue();
      `
    },
    {
      name: 'execute',
      javaCode: `
    Logger logger = new PrefixLogger(new Object[] {
        this.getClass().getSimpleName()
      }, (Logger) x.get("logger"));
    logger.info("execute");

    try {
      while ( true ) {
        foam.nanos.medusa.ClusterConfigSupport support = (foam.nanos.medusa.ClusterConfigSupport) x.get("clusterConfigSupport");
        if ( getEnabled() ) {
          Date now = new Date();

          getCronDAO().where(
                         MLang.AND(
                                   MLang.LTE(Cron.SCHEDULED_TIME, now),
                                   MLang.EQ(Cron.ENABLED, true),
                                   MLang.IN(Cron.STATUS, new ScriptStatus[] {
                                                          ScriptStatus.UNSCHEDULED,
                                                          ScriptStatus.ERROR,
                                     })
                                   )
                         )
            .orderBy(Cron.SCHEDULED_TIME)
            .select(new AbstractSink() {
                             @Override
                             public void put(Object obj, Detachable sub) {
                               Cron cron = (Cron) ((FObject) obj).fclone();
                               PM pm = new PM(this.getClass().getSimpleName(), "cronjob", cron.getId());
                               try {
                                 if ( ! cron.getClusterable() ||
                                      support == null ||
                                      support.cronEnabled(x) ) {
                                   cron.setStatus(ScriptStatus.SCHEDULED);
                                   getCronDAO().put(cron);
                                 }
                               } catch (Throwable t) {
                                 logger.error(this.getClass(), "Error scheduling cron job", cron.getId(), t.getMessage(), t);
                                 pm.error(x, t);
                               } finally {
                                 pm.log(x);
                               }
                             }
                           });
        }
        // Check for new cronjobs every 5 seconds if no current jobs
        // or if their next scheduled execution time is > 5s away
        // Delay at least a little bit to avoid blocking in case of a script error.
        long delay = getCronDelay();
        Date minScheduledTime = getMinScheduledTime();
        if( minScheduledTime != null &&
            getEnabled() ) {
          delay = Math.abs(minScheduledTime.getTime() - System.currentTimeMillis());
          delay = Math.min(getCronDelay(), delay);
          delay = Math.max(500, delay);
        }
        Thread.sleep(delay);
      }
    } catch (Throwable t) {
      logger.error(this.getClass(), t.getMessage());
      ((DAO) x.get("alarmDAO")).put(new foam.nanos.alarming.Alarm.Builder(x)
        .setName(this.getClass().getSimpleName())
        .setSeverity(foam.log.LogLevel.ERROR)
        .setNote(t.getMessage())
        .build());
    }
    `
    }
  ]
});
